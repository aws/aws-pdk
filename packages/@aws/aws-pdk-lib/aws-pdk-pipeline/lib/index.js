"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
__exportStar(require("./pdk-pipeline"), exports);
__exportStar(require("./pdk-pipeline-ts-project"), exports);
__exportStar(require("./pdk-pipeline-py-project"), exports);
__exportStar(require("./pdk-pipeline-java-project"), exports);
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi9zcmMvaW5kZXgudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6Ijs7Ozs7Ozs7Ozs7O0FBQUEsaURBQStCO0FBQy9CLDREQUEwQztBQUMxQyw0REFBMEM7QUFDMUMsOERBQTRDIiwic291cmNlc0NvbnRlbnQiOlsiZXhwb3J0ICogZnJvbSBcIi4vcGRrLXBpcGVsaW5lXCI7XG5leHBvcnQgKiBmcm9tIFwiLi9wZGstcGlwZWxpbmUtdHMtcHJvamVjdFwiO1xuZXhwb3J0ICogZnJvbSBcIi4vcGRrLXBpcGVsaW5lLXB5LXByb2plY3RcIjtcbmV4cG9ydCAqIGZyb20gXCIuL3Bkay1waXBlbGluZS1qYXZhLXByb2plY3RcIjtcbiJdfQ==