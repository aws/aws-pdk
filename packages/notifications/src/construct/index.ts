/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./dynamo-db";
export * from "./sqs";
