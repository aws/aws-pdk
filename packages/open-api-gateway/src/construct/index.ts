/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./open-api-gateway-rest-api";
export * from "./waf/types";
export * from "./authorizers";
export * from "./integrations";
export * from "./spec";
