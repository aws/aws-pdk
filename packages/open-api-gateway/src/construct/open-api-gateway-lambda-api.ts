/*********************************************************************************************************************
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

 Licensed under the Apache License, Version 2.0 (the "License").
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ******************************************************************************************************************** */

import * as path from "path";
import { PDKNag } from "@aws-prototyping-sdk/pdk-nag";
import { CustomResource, Duration, Stack } from "aws-cdk-lib";
import {
  AccessLogFormat,
  ApiDefinition,
  Cors,
  LogGroupLogDestination,
  MethodLoggingLevel,
  RestApiBaseProps,
  SpecRestApi,
} from "aws-cdk-lib/aws-apigateway";
import {
  Effect,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam";
import {
  CfnPermission,
  Code,
  Function as LambdaFunction,
  Runtime,
} from "aws-cdk-lib/aws-lambda";
import { LogGroup } from "aws-cdk-lib/aws-logs";
import { Asset } from "aws-cdk-lib/aws-s3-assets";
import { Provider } from "aws-cdk-lib/custom-resources";
import { NagSuppressions } from "cdk-nag";
import { Construct } from "constructs";
import { Authorizers } from "./authorizers";
import { PrepareApiSpecCustomResourceProperties } from "./prepare-spec-event-handler";
import {
  prepareApiSpec,
  PrepareApiSpecOptions,
} from "./prepare-spec-event-handler/prepare-spec";
import { OpenApiOptions } from "./spec";
import {
  prepareSecuritySchemes,
  serializeAsAuthorizerReference,
} from "./spec/api-gateway-auth";
import { getLabelledFunctions } from "./spec/api-gateway-integrations";
import { functionInvocationUri } from "./spec/utils";

/**
 * Configuration for the OpenApiGatewayLambdaApi construct
 */
export interface OpenApiGatewayLambdaApiProps
  extends RestApiBaseProps,
    OpenApiOptions {
  /**
   * The parsed OpenAPI specification
   */
  readonly spec: any; // Type is OpenAPIV3.Document - however not transpileable via jsii, so we use any.
  /**
   * Path to the JSON open api spec
   */
  readonly specPath: string;
}

/**
 * A construct for creating an api gateway api based on the definition in the OpenAPI spec.
 */
export class OpenApiGatewayLambdaApi extends Construct {
  public readonly api: SpecRestApi;

  constructor(
    scope: Construct,
    id: string,
    props: OpenApiGatewayLambdaApiProps
  ) {
    super(scope, id);

    const {
      integrations,
      spec,
      specPath,
      operationLookup,
      defaultAuthorizer,
      corsOptions,
      ...options
    } = props;

    // Upload the spec to s3 as an asset
    const inputSpecAsset = new Asset(this, "InputSpec", {
      path: specPath,
    });
    // We'll output the prepared spec in the same asset bucket
    const preparedSpecOutputKeyPrefix = `${inputSpecAsset.s3ObjectKey}-prepared`;

    const stack = Stack.of(this);

    const prepareSpecLambdaName = `${PDKNag.getStackPrefix(stack)
      .split("/")
      .join("-")}PrepareSpec`;
    const prepareSpecRole = new Role(this, "PrepareSpecRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        logs: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              resources: [
                `arn:aws:logs:${stack.region}:${stack.account}:log-group:/aws/lambda/${prepareSpecLambdaName}`,
                `arn:aws:logs:${stack.region}:${stack.account}:log-group:/aws/lambda/${prepareSpecLambdaName}:*`,
              ],
            }),
          ],
        }),
        s3: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:getObject"],
              resources: [
                inputSpecAsset.bucket.arnForObjects(inputSpecAsset.s3ObjectKey),
              ],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:putObject"],
              resources: [
                // The output file will include a hash of the prepared spec, which is not known until deploy time since
                // tokens must be resolved
                inputSpecAsset.bucket.arnForObjects(
                  `${preparedSpecOutputKeyPrefix}/*`
                ),
              ],
            }),
          ],
        }),
      },
    });

    NagSuppressions.addResourceSuppressions(
      prepareSpecRole,
      [
        {
          id: "AwsSolutions-IAM5",
          reason:
            "Cloudwatch resources have been scoped down to the LogGroup level, however * is still needed as stream names are created just in time.",
          appliesTo: [
            {
              regex: `/^Resource::arn:aws:logs:${PDKNag.getStackRegionRegex(
                stack
              )}:${PDKNag.getStackAccountRegex(
                stack
              )}:log-group:/aws/lambda/${prepareSpecLambdaName}:\*/g`,
            },
          ],
        },
      ],
      true
    );

    // Create a custom resource for preparing the spec for deployment (adding integrations, authorizers, etc)
    const prepareSpec = new LambdaFunction(this, "PrepareSpec", {
      handler: "index.handler",
      runtime: Runtime.NODEJS_16_X,
      code: Code.fromAsset(
        path.join(__dirname, "../../lib/construct/prepare-spec-event-handler")
      ),
      timeout: Duration.seconds(30),
      role: prepareSpecRole,
      functionName: prepareSpecLambdaName,
    });

    const providerFunctionName = `${prepareSpecLambdaName}-Provider`;
    const providerRole = new Role(this, "PrepareSpecProviderRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        logs: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              resources: [
                `arn:aws:logs:${stack.region}:${stack.account}:log-group:/aws/lambda/${providerFunctionName}`,
                `arn:aws:logs:${stack.region}:${stack.account}:log-group:/aws/lambda/${providerFunctionName}:*`,
              ],
            }),
          ],
        }),
      },
    });

    const provider = new Provider(this, "PrepareSpecProvider", {
      onEventHandler: prepareSpec,
      role: providerRole,
      providerFunctionName,
    });

    NagSuppressions.addResourceSuppressions(
      providerRole,
      [
        {
          id: "AwsSolutions-IAM5",
          reason:
            "Cloudwatch resources have been scoped down to the LogGroup level, however * is still needed as stream names are created just in time.",
        },
      ],
      true
    );
    NagSuppressions.addResourceSuppressions(
      provider,
      [
        {
          id: "AwsSolutions-L1",
          reason:
            "Latest runtime cannot be configured. CDK will need to upgrade the Provider construct accordingly.",
        },
      ],
      true
    );

    const prepareSpecOptions: PrepareApiSpecOptions = {
      integrations: Object.fromEntries(
        Object.entries(integrations).map(([operation, integration]) => [
          operation,
          {
            functionInvocationUri: functionInvocationUri(
              this,
              integration.function
            ),
            methodAuthorizer: serializeAsAuthorizerReference(
              integration.authorizer || defaultAuthorizer || Authorizers.none()
            ),
          },
        ])
      ),
      securitySchemes: prepareSecuritySchemes(
        this,
        integrations,
        defaultAuthorizer
      ),
      corsOptions: corsOptions && {
        allowHeaders: corsOptions.allowHeaders || Cors.DEFAULT_HEADERS,
        allowMethods: corsOptions.allowMethods || Cors.ALL_METHODS,
        allowOrigins: corsOptions.allowOrigins,
        statusCode: corsOptions.statusCode || 204,
      },
      operationLookup,
    };

    // Spec preparation will happen in a custom resource lambda so that references to lambda integrations etc can be
    // resolved. However, we also prepare inline to perform some additional validation at synth time.
    const preparedSpec = prepareApiSpec(spec, prepareSpecOptions);

    const prepareApiSpecCustomResourceProperties: PrepareApiSpecCustomResourceProperties =
      {
        inputSpecLocation: {
          bucket: inputSpecAsset.bucket.bucketName,
          key: inputSpecAsset.s3ObjectKey,
        },
        outputSpecLocation: {
          bucket: inputSpecAsset.bucket.bucketName,
          key: preparedSpecOutputKeyPrefix,
        },
        ...prepareSpecOptions,
      };

    const prepareSpecCustomResource = new CustomResource(
      this,
      "PrepareSpecCustomResource",
      {
        serviceToken: provider.serviceToken,
        properties: prepareApiSpecCustomResourceProperties,
      }
    );

    // Create the api gateway resources from the spec, augmenting the spec with the properties specific to api gateway
    // such as integrations or auth types
    this.api = new SpecRestApi(this, id, {
      apiDefinition: ApiDefinition.fromBucket(
        inputSpecAsset.bucket,
        prepareSpecCustomResource.getAttString("outputSpecKey")
      ),
      deployOptions: {
        accessLogDestination: new LogGroupLogDestination(
          new LogGroup(scope, `AccessLogs`)
        ),
        accessLogFormat: AccessLogFormat.clf(),
        loggingLevel: MethodLoggingLevel.INFO,
      },
      ...options,
    });

    this.api.node.addDependency(prepareSpecCustomResource);

    // While the api will be updated when the output path from the custom resource changes, CDK still needs to know when
    // to redeploy the api. This is achieved by including a hash of the spec in the logical id (internalised in the
    // addToLogicalId method since this is how changes of individual resources/methods etc trigger redeployments in CDK)
    this.api.latestDeployment?.addToLogicalId(preparedSpec);

    // Grant API Gateway permission to invoke each lambda which implements an integration or custom authorizer
    getLabelledFunctions(props).forEach(({ label, function: lambda }) => {
      new CfnPermission(this, `LambdaPermission-${label}`, {
        action: "lambda:InvokeFunction",
        principal: "apigateway.amazonaws.com",
        functionName: lambda.functionArn,
        sourceArn: stack.formatArn({
          service: "execute-api",
          resource: this.api.restApiId,
          resourceName: "*/*",
        }),
      });
    });

    NagSuppressions.addResourceSuppressions(
      this,
      [
        {
          id: "AwsSolutions-IAM4",
          reason:
            "Cloudwatch Role requires access to create/read groups at the root level.",
          appliesTo: [
            {
              regex: `/^Policy::arn:${PDKNag.getStackPartitionRegex(
                stack
              )}:iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs$/g`,
            },
          ],
        },
      ],
      true
    );

    NagSuppressions.addResourceSuppressions(
      this,
      [
        {
          id: "AwsSolutions-APIG2",
          reason:
            "This construct implements fine grained validation via OpenApi.",
        },
      ],
      true
    );
  }
}
