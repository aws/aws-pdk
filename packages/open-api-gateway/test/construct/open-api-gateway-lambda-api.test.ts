/*********************************************************************************************************************
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

 Licensed under the Apache License, Version 2.0 (the "License").
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ******************************************************************************************************************** */

import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { AuthorizationType, Cors } from "aws-cdk-lib/aws-apigateway";
import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
import { OpenAPIV3 } from "openapi-types";
import { OpenApiGatewayLambdaApi } from "../../lib/construct";

const sampleSpec: OpenAPIV3.Document = {
  openapi: "3.0.3",
  info: {
    version: "1.0.0",
    title: "Test API",
  },
  paths: {
    "/test": {
      get: {
        operationId: "testOperation",
        responses: {
          200: {
            description: "Successful response",
            content: {
              "application/json": {
                schema: {
                  type: "object",
                  properties: {
                    message: {
                      type: "string",
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
};

const operationLookup = {
  testOperation: {
    method: "get",
    path: "/test",
  },
};

describe("OpenAPI Gateway Lambda Api Construct Unit Tests", () => {
  it("Synth", () => {
    const stack = new Stack();
    new OpenApiGatewayLambdaApi(stack, "ApiTest", {
      spec: sampleSpec,
      operationLookup,
      integrations: {
        testOperation: {
          function: new Function(stack, "Lambda", {
            code: Code.fromInline("code"),
            handler: "handler",
            runtime: Runtime.NODEJS_16_X,
          }),
        },
      },
    });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("With IAM Auth and CORS", () => {
    const stack = new Stack();
    new OpenApiGatewayLambdaApi(stack, "ApiTest", {
      authType: AuthorizationType.IAM,
      corsOptions: {
        allowOrigins: Cors.ALL_ORIGINS,
        allowMethods: Cors.ALL_METHODS,
        allowCredentials: true,
        statusCode: 200,
      },
      spec: sampleSpec,
      operationLookup,
      integrations: {
        testOperation: {
          function: new Function(stack, "Lambda", {
            code: Code.fromInline("code"),
            handler: "handler",
            runtime: Runtime.NODEJS_16_X,
          }),
        },
      },
    });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("Should Throw When Integration Missing", () => {
    const stack = new Stack();
    expect(
      () =>
        new OpenApiGatewayLambdaApi(stack, "ApiTest", {
          spec: sampleSpec,
          operationLookup,
          integrations: {},
        })
    ).toThrow(
      "Missing required integration for operation testOperation (get /test)"
    );
  });
});
