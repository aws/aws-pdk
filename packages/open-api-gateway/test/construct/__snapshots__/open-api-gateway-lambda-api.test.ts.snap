// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI Gateway Lambda Api Construct Unit Tests Synth 1`] = `
Object {
  "Outputs": Object {
    "ApiTestEndpointC91A82E8": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "ApiTest0B514494",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "ApiTestDeploymentStageprod77F35FA8",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ApiTest0B514494": Object {
      "Properties": Object {
        "Body": Object {
          "info": Object {
            "title": "Test API",
            "version": "1.0.0",
          },
          "openapi": "3.0.3",
          "paths": Object {
            "/test": Object {
              "get": Object {
                "operationId": "testOperation",
                "responses": Object {
                  "200": Object {
                    "content": Object {
                      "application/json": Object {
                        "schema": Object {
                          "properties": Object {
                            "message": Object {
                              "type": "string",
                            },
                          },
                          "type": "object",
                        },
                      },
                    },
                    "description": "Successful response",
                    "headers": Object {},
                  },
                },
                "x-amazon-apigateway-auth": Object {
                  "type": "NONE",
                },
                "x-amazon-apigateway-integration": Object {
                  "httpMethod": "POST",
                  "passthroughBehavior": "WHEN_NO_MATCH",
                  "type": "AWS_PROXY",
                  "uri": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":apigateway:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":lambda:path/2015-03-31/functions/",
                        Object {
                          "Fn::GetAtt": Array [
                            "LambdaD247545B",
                            "Arn",
                          ],
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
            },
          },
          "x-amazon-apigateway-gateway-responses": Object {
            "BAD_REQUEST_BODY": Object {
              "responseTemplates": Object {
                "application/json": "{\\"message\\": \\"$context.error.validationErrorString\\"}",
              },
              "statusCode": 400,
            },
          },
          "x-amazon-apigateway-request-validator": "all",
          "x-amazon-apigateway-request-validators": Object {
            "all": Object {
              "validateRequestBody": true,
              "validateRequestParameters": true,
            },
          },
        },
        "Name": "ApiTest",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ApiTestAccount2AC853EB": Object {
      "DependsOn": Array [
        "ApiTest0B514494",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "ApiTestCloudWatchRole49F4F360",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "ApiTestCloudWatchRole49F4F360": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiTestDeployment59A5FA032c19e7dfc4d878c4b5d945b82a5312c1": Object {
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "ApiTest0B514494",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ApiTestDeploymentStageprod77F35FA8": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "ApiTestDeployment59A5FA032c19e7dfc4d878c4b5d945b82a5312c1",
        },
        "RestApiId": Object {
          "Ref": "ApiTest0B514494",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ApiTestLambdaPermissiontestOperationECAC1A2D": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Ref": "LambdaD247545B",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiTest0B514494",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "LambdaD247545B": Object {
      "DependsOn": Array [
        "LambdaServiceRoleA8ED4D3B",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "code",
        },
        "Handler": "handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaServiceRoleA8ED4D3B",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaServiceRoleA8ED4D3B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`OpenAPI Gateway Lambda Api Construct Unit Tests With IAM Auth and CORS 1`] = `
Object {
  "Outputs": Object {
    "ApiTestEndpointC91A82E8": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "ApiTest0B514494",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "ApiTestDeploymentStageprod77F35FA8",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ApiTest0B514494": Object {
      "Properties": Object {
        "Body": Object {
          "info": Object {
            "title": "Test API",
            "version": "1.0.0",
          },
          "openapi": "3.0.3",
          "paths": Object {
            "/test": Object {
              "get": Object {
                "operationId": "testOperation",
                "responses": Object {
                  "200": Object {
                    "content": Object {
                      "application/json": Object {
                        "schema": Object {
                          "properties": Object {
                            "message": Object {
                              "type": "string",
                            },
                          },
                          "type": "object",
                        },
                      },
                    },
                    "description": "Successful response",
                    "headers": Object {
                      "Access-Control-Allow-Headers": Object {
                        "schema": Object {
                          "type": "string",
                        },
                      },
                      "Access-Control-Allow-Methods": Object {
                        "schema": Object {
                          "type": "string",
                        },
                      },
                      "Access-Control-Allow-Origin": Object {
                        "schema": Object {
                          "type": "string",
                        },
                      },
                    },
                  },
                },
                "x-amazon-apigateway-auth": Object {
                  "type": "AWS_IAM",
                },
                "x-amazon-apigateway-integration": Object {
                  "httpMethod": "POST",
                  "passthroughBehavior": "WHEN_NO_MATCH",
                  "type": "AWS_PROXY",
                  "uri": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":apigateway:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":lambda:path/2015-03-31/functions/",
                        Object {
                          "Fn::GetAtt": Array [
                            "LambdaD247545B",
                            "Arn",
                          ],
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
              "options": Object {
                "description": "Enable CORS by returning the correct headers",
                "responses": Object {
                  "200": Object {
                    "content": Object {},
                    "description": "Default response for CORS method",
                    "headers": Object {
                      "Access-Control-Allow-Headers": Object {
                        "schema": Object {
                          "type": "string",
                        },
                      },
                      "Access-Control-Allow-Methods": Object {
                        "schema": Object {
                          "type": "string",
                        },
                      },
                      "Access-Control-Allow-Origin": Object {
                        "schema": Object {
                          "type": "string",
                        },
                      },
                    },
                  },
                },
                "summary": "CORS Support",
                "x-amazon-apigateway-integration": Object {
                  "requestTemplates": Object {
                    "application/json": "{\\"statusCode\\": 200}",
                  },
                  "responses": Object {
                    "default": Object {
                      "responseParameters": Object {
                        "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                        "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                        "method.response.header.Access-Control-Allow-Origin": "'*'",
                      },
                      "responseTemplates": Object {
                        "application/json": "{}",
                      },
                      "statusCode": "200",
                    },
                  },
                  "type": "mock",
                },
              },
            },
          },
          "x-amazon-apigateway-gateway-responses": Object {
            "BAD_REQUEST_BODY": Object {
              "responseParameters": Object {
                "gatewayresponse.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                "gatewayresponse.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                "gatewayresponse.header.Access-Control-Allow-Origin": "'*'",
              },
              "responseTemplates": Object {
                "application/json": "{\\"message\\": \\"$context.error.validationErrorString\\"}",
              },
              "statusCode": 400,
            },
          },
          "x-amazon-apigateway-request-validator": "all",
          "x-amazon-apigateway-request-validators": Object {
            "all": Object {
              "validateRequestBody": true,
              "validateRequestParameters": true,
            },
          },
        },
        "Name": "ApiTest",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ApiTestAccount2AC853EB": Object {
      "DependsOn": Array [
        "ApiTest0B514494",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "ApiTestCloudWatchRole49F4F360",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "ApiTestCloudWatchRole49F4F360": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiTestDeployment59A5FA030122928b0149760ff9dad9a1e9321b54": Object {
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "ApiTest0B514494",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ApiTestDeploymentStageprod77F35FA8": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "ApiTestDeployment59A5FA030122928b0149760ff9dad9a1e9321b54",
        },
        "RestApiId": Object {
          "Ref": "ApiTest0B514494",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ApiTestLambdaPermissiontestOperationECAC1A2D": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Ref": "LambdaD247545B",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiTest0B514494",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "LambdaD247545B": Object {
      "DependsOn": Array [
        "LambdaServiceRoleA8ED4D3B",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "code",
        },
        "Handler": "handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaServiceRoleA8ED4D3B",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaServiceRoleA8ED4D3B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
